lib/src/task_either.dart

```diff
@@ -26,6 +40,13 @@ class TaskEither<L, R> extends HKT2<_TaskEitherHKT, L, R>
   /// Build a [TaskEither] from a function returning a `Future<Either<L, R>>`.
   const TaskEither(this._run);
 
+  /// Initialize a **Do Notation** chain.
+  // ignore: non_constant_identifier_names
+  factory TaskEither.Do(DoFunctionTaskEither<L, R> f) => TaskEither.tryCatch(
+        () => f(_doAdapter<L>()),
+        (error, _) => (error as _TaskEitherThrow<L>).value,
```

Comment:
Similar to Option, I think you will want to put a conditional here, and rethrow any errors that are not `_TaskEitherThrow`.

```dart
error is _TaskEitherThrow<L> ? error.value : throw error
```